src/bosh-director/lib/bosh/director/problem_resolver.rb

```diff
@@ -52,18 +55,80 @@ def apply_resolutions(resolutions)
 
     private
 
-    def process_problem(problem)
-      if problem.state != 'open'
-        reason = "state is '#{problem.state}'"
-        track_and_log("Ignoring problem #{problem.id} (#{reason})")
-      elsif problem.deployment_id != @deployment.id
-        reason = 'not a part of this deployment'
-        track_and_log("Ignoring problem #{problem.id} (#{reason})")
+    def parallel_each(num_threads, ary)
+      if num_threads > 1
+        ThreadPool.new(max_threads: num_threads).wrap do |pool|
+          ary.each do |entry|
+            pool.process do
+              yield entry
+            end
+          end
+        end
       else
+        ary.each do |entry|
+          yield entry
+        end
+      end
+    end
+
+    def process_problem(problem)
+      if problem.open?
         apply_resolution(problem)
+      else
+        track_and_log("Ignoring problem #{problem.id} (state is '#{problem.state}')")
       end
     end
 
+    def filter_instance_groups(igs_to_problems)
+      ig_partition_to_ig_names_with_problems = {}
+      BatchMultiInstanceGroupUpdater.partition_jobs_by_serial(@instance_groups).each do |ig_partition|
+        igs_with_problems = []
+        ig_partition.each { |ig| igs_with_problems << ig.name if igs_to_problems.key?(ig.name) }
```

Comment:
A minor nit, but these blocks are harder to read when using `each` instead of the more idiomatic ruby array methods like `map` or `select`/`reject` or even `reduce`. For instance, theses two lines could be replaced with
```rb
igs_with_problems = ig_partition.map(&:name).select { |ig_name| igs_to_problems.key?(ig_name) }
```
This is much more readable to me. There are similar patterns throughout the PR (most of the occurrences of `each`). I wouldn't want to block the PR on it, but would still appreciate if you could make them more idiomatic ruby.

Reply:
It also feels much cleaner to have a proper return value via `map` instead of filling an array via a side effect from within a loop.